accelerate-llvm-ptx/src/Data/Array/Accelerate/LLVM/PTX/CodeGen/Base.hs

```diff
@@ -325,6 +331,180 @@ atomicAdd_f t addr val =
+      IntegralNumType TypeWord64 -> shfl_64_bit a
+      FloatingNumType TypeDouble -> shfl_64_bit a
+
+    VectorScalarType vectyp -> shfl_vec vectyp a
+  where
+    shfl_i32 :: Operands Int32 -> CodeGen PTX (Operands Int32)
+    shfl_i32 x = mk_shfl mode "i32" x delta
+
+    shfl_f32 :: Operands Float -> CodeGen PTX (Operands Float)
+    shfl_f32 x = mk_shfl mode "f32" x delta
+
+    -- Shuffle a (<32) bit integral
+    cast_shfl_cast_i :: (IsIntegral a, IsNum a) => Operands a -> CodeGen PTX (Operands a)
+    cast_shfl_cast_i = A.fromIntegral integralType numType
+                    >=> shfl_i32
+                    >=> A.fromIntegral integralType numType
+
+    -- Shuffle a 16 bit floating point number
+    cast_shfl_cast_f16 :: Operands Half -> CodeGen PTX (Operands Half)
+    cast_shfl_cast_f16 = bitcast scalarType scalarType
+                      >=> cast_shfl_cast_i @Int16
+                      >=> bitcast scalarType scalarType
+
+    -- Shuffle a 64 bit thing
+    shfl_64_bit :: IsScalar a => Operands a -> CodeGen PTX (Operands a)
+    shfl_64_bit x = do
+      let vectype = VectorScalarType (VectorType 2 (NumSingleType (IntegralNumType TypeInt32))) :: ScalarType (Vec 2 Int32)
+      x' <- bitcast scalarType vectype x
+      x'' <- shfl mode (TupRsingle vectype) x' delta
+      bitcast vectype scalarType x''
+
+    shfl_vec :: VectorType a -> Operands a -> CodeGen PTX (Operands a)
+    shfl_vec (VectorType 0 _) x = return x
+    shfl_vec (VectorType n t) x = go (P.fromIntegral n) x (VectorType n t)
+
+    go :: Int32 -> Operands (Vec n a) -> VectorType (Vec n a) -> CodeGen PTX (Operands (Vec n a))
+    go 0 x _ = return x
+    go n x vt@(VectorType _ t) = do
+      e <- instr $ ExtractElement (n-1) (op vt x)
+      e' <- shfl mode (TupRsingle (SingleScalarType t)) e delta
+      x' <- instr $ InsertElement (n-1) (op vt x) (op (SingleScalarType t) e')
+      go (n-1) x' vt
+
+
+-- Only works for 32-bit int or floats!!
+mk_shfl :: (IsPrim a)
+        => Label -- direction: ["up","down","bfly","idx"] (the latter two probably represent the "xor" and "get from an index" variants)
+        -> Label -- the type, "i32" or "f32"
+        -> Operands a                  -- value to give
+        -> Operands Word32             -- delta
+        -> CodeGen PTX (Operands a)   -- value received
+mk_shfl mode typ val delta = do
+
+  -- In CUDA, the default for the final parameter, width, is warpsize.
+  -- Setting the width lower (always power of 2) emulates the shfl behaviour with a smaller warpsize.
+  -- Behind the scenes, a bunch of instructions happen with this width parameter
+  -- before they get passed into the actual shfl instruction. Here, we have to
+  -- directly set them to the 'actual' width parameter. Below, the formula that
+  -- clang compiles to is in the comments, but we just use a constant rather
+  -- than recompute these every time a shfl is invoked (which is what clang-cuda does).
+  --
+  -- -DB
+
+  let width = case mode of
```

Comment:
But this is unforgivable... ._.

Reply:
On the bright side: Unless this gets optimised out later, we now have more efficient shfl instructions than clang does :)